task_processing/plugins/mesos/logging_executor.py

```diff
@@ -0,0 +1,151 @@
+        self.downstream_executor = downstream_executor
+
+        self.src_queue = downstream_executor.get_event_queue()
+        self.dest_queue = Queue()
+        self.stopping = False
+
+        # No lock is needed because log_metadata is touched in a single thread
+        self.log_metadata = m()
+
+        self.logging_thread = Thread(target=self.logging_loop)
+        self.logging_thread.daemon = True
+        self.logging_thread.start()
+
+    def build_task_log_url(self, task_log_url, event):
+        try:
+            response = json.load(urllib.request.urlopen(
+                task_log_url + 'files/debug'))
+        except Exception as e:
+            log.error("Failed to fetch files {error}".format(error=e))
+            return ''
+
+        for key in response.keys():
+            if event.raw.executor_id.value in key and \
+                    event.raw.container_status.container_id.value in key:
+                return (task_log_url + 'files/read?path=' + key + '/')
+        return ''
+
+    def stream_task_log(self, task_id):
+        log_md = self.log_metadata[task_id]
+        # Do not fetch logs if url hasn't been built
+        if log_md.log_url is '':
+            return
+
+        offsets = {
+            'stdout': log_md.log_offsets['stdout'],
+            'stderr': log_md.log_offsets['stderr']
+        }
+        for f in ['stdout', 'stderr']:
+            offset = offsets[f]
+            while True:
+                url = log_md.log_url + f + \
+                    '&length=' + str(TASK_LOG_CHUNK_LEN) + \
+                    '&offset=' + str(offset)
+                try:
+                    response = json.load(urllib.request.urlopen(url))
+                    log_length = len(response['data'])
+                    for line in response['data'].splitlines():
+                        print(
+                            task_id + ": " + line,
+                            file=sys.stderr if f is 'stderr' else sys.stdout
+                        )
+                    offset = offset + log_length
+                    # Stop if there is no more data
+                    if log_length < TASK_LOG_CHUNK_LEN:
+                        break
+                except Exception as e:
+                    log.error("Failed to fetch task {f} log {error}".format(
+                        f=f, error=e
+                    ))
+            # Update offset of this stream
+            offsets[f] = offset
+
+        # Update both offsets for the task
+        self.log_metadata = self.log_metadata.set(
+            task_id,
+            log_md.set(log_offsets=pmap(offsets)),
+        )
+
+    def logging_loop(self):
+        while True:
+            # process downstream events
+            while not self.src_queue.empty():
+                e = self.src_queue.get()
+                self.dest_queue.put(e)
+                # Record the 1st part of the log url
+                if e.kind == 'task' and e.platform_type == 'staging':
```

Comment:
do you need to have separate processing at staging vs running? couldn't you do all of the url creation when you get the first TASK_RUNNING, and ignore any further events you get for the same task id?

Reply:
The staging event passes up the agent address (offer). This info will not repeat in the running event.